### PR TITLE
Fix MessageIndex determinism

### LINKED ISSUE #201: Fix MessageIndex determinism
## Description

values stored in `MessageIndex` should be ordered by `message_id`, currently they are not

## Motivation

required for resolution

## To-do list

- [x] Ensure values added to `MessageIndex` are sorted by `T::message_id`
  - [x] Replace `scope.insert` with equivalent binary search

## Change checklist

- [ ] The feature or fix is implemented in Rust and across all bindings whereas possible.
- [ ] The feature or fix has sufficient testing coverage
- [ ] All tests and examples build and run locally as expected
- [ ] Every piece of code has been document according to the documentation guidelines.
- [ ] If conceptual documentation (mdbook) and examples highlighting the feature exist, they are properly updated.
- [ ] If the feature is not currently documented, a documentation task Issue has been opened to address this.

### PR DESCRIPTION
**Pull request by DeppLearning**
# Description of change
- Multiple elements with the same previous_message_id are ordered by their message_id
-  
## Links to any relevant issues
- fixes #201
- 
## Type of change
Add an `x` to the boxes that are relevant to your changes.

- [x] Bug fix (a non-breaking change which fixes an issue)

## How the change has been tested
Describe the tests that you ran to verify your changes.
Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist
Add an `x` to the boxes that are relevant to your changes.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: identity-iota/src/tangle/message_index.rs [hunk 1]
```diff
@@ -47,9 +47,13 @@ where
 {
   pub fn insert(&mut self, element: T) {
     let key: &MessageId = element.previous_message_id();
-
+    let msg_id: &MessageId = element.message_id();
```
Comment (l1h3r) [inline identity-iota/src/tangle/message_index.rs:50]: This could go in the `if` branch

**Review by l1h3r — APPROVED** [review]
🚀

**Comment by JelleMillenaar** [comment]
I see the ordering is fixed, does it also actually use the ordering to only apply the first VALID message in the ordered list to the chain, ignoring the rest?

**Comment by l1h3r** [comment]
> I see the ordering is fixed, does it also actually use the ordering to only apply the first VALID message in the ordered list to the chain, ignoring the rest?

Yeah, all messages that fail serialization/validation are ignored and we only use one message per id

**Review by l1h3r — APPROVED** [review]

**Pull request merged**